R/as-tibble.R

```diff
@@ -1,26 +1,77 @@
+#' @importFrom crunch getDimTypes
+#' @importFrom dplyr bind_cols
+#' @importFrom purrr map2 map reduce
+#' @importFrom stringr str_extract
 as_tibble.CrunchCube <- function (x, ...) {
     ## TODO: Consider using `dplyr::tbl_cube` class
-    ## TODO: NA behavior? Add way to get missings not missing. (This uses
-    ## as.array to remove extraneous dims, and doing so applies the '@useNA'
-    ## from the cube)
-    ## TODO: deal with naming for array (and NUMR) variable dims (var.categories?)
+    
+    dnames <- dimnames(x@arrays$.unweighted_counts)
+    measures <- names(x@arrays)
 
-    ## 1) get the measures
-    ## TODO: better access methods for secondary measures in Cube
-    ## TODO: if weighted query, include ".unweighted_counts"
-    measure_names <- setdiff(names(x@arrays), ".unweighted_counts")
-    vals <- sapply(measure_names, function (m) {
-        as.vector(crunch:::cubeToArray(x, m))
-    }, simplify=FALSE)
+    # Cubes can have multiple measures, which are represented as their own column
+    # in the tibble.
+    measure_vals <- sapply(measures, function(y) {
+        as.vector(x@arrays[[y]])
+    }, simplify = FALSE)
+    
+    names(measure_vals)[names(measure_vals) == ".unweighted_counts"] <- "row_count"
 
-    ## 2) If there are dimnames, expand.grid and c(bind) them
-    dnames <- dimnames(as.array(x))
-    ## NULL if scalar value, which means no "group_by"
+    # If there are dimnames, expand.grid and bind them. We also change the names
+    # of the two array dimensions to avoid duplicated variable names in  the
+    # tibble.
     if (!is.null(dnames)) {
-        dims <- do.call(expand.grid, dnames)
-        vals <- c(dims, vals)
+        types <- getDimTypes(x)
+
+        # Change MR selection vars to T/F/NA
+        is_selected <- types == "mr_selections"
+        dnames <- map2(dnames, is_selected, ~{
+            if (.y) {
+                return(c(TRUE, FALSE, NA))
+            } else {
+                return (.x)
+            }
+        })
+        suffixes <- str_extract(types, "_.*$")
+        is_array_var <- !is.na(suffixes)
+        names(dnames)[is_array_var] <- paste0(names(dnames)[is_array_var], suffixes[is_array_var])
+        out <- do.call(expand.grid, dnames)
+        names(out) <- add_duplicate_suffix(names(out))
+        
+        # Identify which elements of cube represent missing values
+        out$is_missing <- x@dims %>%
+            map("missing") %>%
+            expand.grid() %>%
+            reduce(`|`)
+        out <- bind_cols(out, measure_vals)
+    } else {
+        # scalar values, which means no group_by
+        out <- bind_cols(measure_vals)
     }
-    return(as_tibble(vals))
+    return(as_tibble(out, ... ))
 }
+
+
+#' @importFrom purrr walk
+add_duplicate_suffix <- function(names, sep = "_"){
+    u_names <- unique(names)
+    walk(u_names, ~{
+        dupes <- names == .
+        if (sum(dupes) != 1) {
+            names[dupes] <<- paste0(names[dupes], sep, 1:sum(dupes))
```

Comment:
The global assignment here gives me a little pause — is it possible to do this without it, or at least a bit more constrained than global assignment?

Reply:
Is there something you're worried about in particular? The thought was that it's super-assigned to the function's argument so I don't think it can affect state outside of the function. It's possible to do the same thing with a for loop, but the reason I prefer the `walk` pattern is that it makes it clear that you're calling the function for its side effects, and it's also clear which part of the function is impure.  I can't think of a case where the loop would generate a different value from `walk + <<-`